pkg/types/upkeep_config.go

```diff
@@ -0,0 +1,86 @@
+package types
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/fxamacker/cbor/v2"
+)
+
+// UpkeepConfig is the interface for all upkeep configs.
+// It is used to encode and decode the config into bytes,
+// and validate the config.
+type UpkeepConfig interface {
+	// validates the config values
+	Validate() error
+	// encodes the config into bytes
+	Encode() ([]byte, error)
+	// decodes the config from bytes
+	Decode(raw []byte) error
+}
+
+// TODO: implement ConditionalUpkeepConfig
+
+// LogUpkeepConfig holds the settings for a log upkeep
+type LogUpkeepConfig struct {
+	// Address is required, contract address w/ 0x prefix
+	Address string `cbor:"a"`
+	// Topic is required, 32 bytes, w/o 0x prefixed
+	Topic string `cbor:"sig"`
+	// Filter1 is optional, needs to be left-padded to 32 bytes
+	Filter1 string `cbor:"f1,omitempty"`
+	// Filter2 is optional, needs to be left-padded to 32 bytes
+	Filter2 string `cbor:"f2,omitempty"`
+	// Filter3 is wildcard filter if missing
+	Filter3 string `cbor:"f3,omitempty"`
+}
+
+var _ UpkeepConfig = &LogUpkeepConfig{}
+
+var (
+	ErrContractAddrIsMissing = errors.New("missing required field: contract address")
+	ErrTopicIsMissing        = errors.New("missing required field: topic")
+	ErrTopicPrefix           = errors.New("invalid topic: prefixed with 0x")
+)
+
+func (cfg *LogUpkeepConfig) Validate() error {
+	cfg.defaults()
+
+	if len(cfg.Address) == 0 {
+		return ErrContractAddrIsMissing
+	}
+	if n := len(cfg.Topic); n == 0 {
```

Comment:
🏄 If `n` isn't being used this can just be `if len(cfg.Topic) == 0`?

Reply:
length check was extracted to its own function